pandas/core/internals/concat.py

```diff
@@ -283,8 +289,16 @@ def _concat_managers_axis0(
             nb._mgr_locs = nb._mgr_locs.add(offset)
             blocks.append(nb)
 
+        if not made_copy and not copy and using_copy_on_write():
+            refs.extend([weakref.ref(blk) for blk in mgr.blocks])
+            parent = mgr
```

Comment:
We are obviously overwriting parent here when we have more than one object, is this a problem?

Reply:
Ah, yes, I suppose that is a problem. `parent` is just used to keep a reference but is not otherwise used, so it doesn't really matter what object it is. So I think we can also make it a list of parent objects in this case.

(will try to make a test that fails because of this; generally it is for chained operations, which you might have less often with concat)

Reply:
A "chained" concat (not very normal usage .., but still should work correctly ;)):

```
In [9]: pd.options.mode.copy_on_write = True

In [36]: res = pd.concat([pd.concat([df1, df2], axis=1), df3], axis=1)

In [37]: df1 = pd.DataFrame({'a': [1, 2, 3], 'b': [0.1, 0.2, 0.3]})

In [38]: df2 = pd.DataFrame({'c': [4, 5, 6]})

In [39]: df3 = pd.DataFrame({'d': [4, 5, 6]})

In [40]: res = pd.concat([pd.concat([df1, df2], axis=1), df3], axis=1)

In [41]: df1._mgr._has_no_reference(0)
Out[41]: True

In [42]: df1.iloc[0, 0] = 100

In [43]: res
Out[43]: 
     a    b  c  d
0  100  0.1  4  4
1    2  0.2  5  5
2    3  0.3  6  6

```

Reply:
Yep, definitely agree. I'll keep track of a list of parent objects then. Series does not work correctly either, but this should be straightforward to fix I think